File: server/Models/Livestock.cs
namespace Server.Models;

public class Livestock
{
    public int Id { get; set; }

    /// <summary>A <see cref="LivestockKind"/> name — an open set of built-in defaults plus
    /// whatever custom kinds the user added, referenced by name like Stock.Type.</summary>
    public string Type { get; set; } = string.Empty;
    public int Count { get; set; }
    public string Name { get; set; } = string.Empty;
    public int FarmId { get; set; }

    /// <summary>
    /// What this group produces — the <see cref="ProductionType"/> every one of its
    /// <see cref="AnimalProduction"/> records is collected under. Declared on the group rather than
    /// chosen per record: a herd yields what it yields, and choosing again on each batch is how one
    /// group's history ends up split across unrelated outputs. Settled once set, since the records
    /// already collected are counted under it.
    ///
    /// Null only for a group that predates this and has collected nothing (one with records was
    /// given the type those records were collected under); such a group declares what it produces
    /// before its first record.
    /// </summary>
    public int? ProductionTypeId { get; set; }

    /// <summary>
    /// This group's meat — the <see cref="ProductionType"/> a realization of it is recorded under,
    /// kept apart from <see cref="ProductionTypeId"/> because a herd that yields milk still has
    /// meat. One type per group, named after it, rather than a single shared "Meat", so one
    /// herd's meat is told apart from another's on the balances and in the marketplace.
    ///
    /// Held as an id rather than found by name when needed: the row is named in the language of
    /// whoever created the group, so a farm working in two languages has no name to compute.
    /// Null for a group that has never been realized, including every group predating this.
    /// </summary>
    public int? MeatProductionTypeId { get; set; }
}

File: web/src/types/livestock.ts
// An animal type is now an open set — the built-in defaults plus whatever custom LivestockKinds
// a user has added (see services/livestock-kind-service.ts) — so it's a plain string, not a union.
export type AnimalType = string;

export type Livestock = {
  id: number;
  type: AnimalType;
  count: number;
  name: string;
  farmId: number;
  /** The first of {@link productionTypeIds}, kept for older clients that know only one output.
   *  Null for a group that declares none. */
  productionTypeId: number | null;
  /** Everything the group produces — every record it collects is filed under one of these. An
   *  output can be added at any time; one the group has already collected under cannot be dropped
   *  (the server answers 409). Empty for a group that declares none. */
  productionTypeIds: number[];
  /** This group's meat — the type a realization of it is recorded under, separate from what it
   *  produces day to day. Created with the group and named after its animal, so rabbit meat and
   *  cow meat are not the same output. Null for a group recorded before this existed. */
  meatProductionTypeId: number | null;
  // No realized-or-not here: realization is one animal's, and which of a group's animals have been
  // realized is told by its records — see AnimalProduction.isRealization.
  /** Whether the group has been removed from the livestock page. Removing marks rather than drops,
   *  so everything collected from it still reads back; it is left out of the list and of the
   *  balance until the removed holdings are asked for. */
  isDeleted: boolean;
};

export type LivestockInput = Omit<Livestock, 'id' | 'isDeleted'>;
